File: tracer_version1/src/geometry.jl
module geometry
    importall vec3
    import Base: intersect
    import ray: Ray
    export Geometry, Sphere, Plane,Triangle, intersect
    abstract Geometry

    immutable Sphere <: Geometry
        r::Float64
        center::Vec3f
    end
    """
    norm = (A,B,C)
    |norm| = 1 and does not checked
    norm*x+dist=0 as the equation
    """
    immutable Plane <: Geometry
        norm::Vec3f
        dist::Float64
    end

    immutable Triangle <: Geometry
        p0::Vec3f
        p1::Vec3f
        p2::Vec3f
    end

    """
    solve, return the -Inf,+Inf for complex solution
    """
    function solve_quad(a, b, c)
        delta = b^2-4a*c
        if delta<0
            return -Inf, Inf
        end
        sqrt_delta = sqrt(delta)
        return (-b-sqrt_delta)/2a, (-b+sqrt_delta)/2a
    end
    """
    intersect with the sphere, it should result to two point (real or imaginary)
    however, we will only interested in the nearest positive point,
    if point does not exists, return +infinity
    """
    function intersect(sphere::Sphere, ray::Ray)
        println("tested")
        a = dot(ray.dir, ray.dir)
        o_c = ray.origin-sphere.center
        b = 2.0*dot(o_c, ray.dir)
        c = dot(o_c, o_c)
        dist1, dist2 = solve_quad(a, b, c)
        if dist1>0
            return dist1
        elseif dist2>0
            return dist2
        else
            return Inf
        end
    end

    """
    intersect with the plane, it will return the nearest positive point or +infinity
    """
    function intersect(plane::Plane, ray::Ray)
        dist = (plane.dist-dot(ray.origin, plane.norm))/dot(ray.origin, plane.norm)
        return dist>0 ? dist : Inf
    end


    """
    test if the result is inside the triangle
    It is also a Barycentric Technique,
    found on the web
    http://www.blackpawn.com/texts/pointinpoly/
    """
    function inside_triangle(t::Triangle, p::Vec3f)
        # Compute vectors
        v0 = t.p2-t.p0
        v1 = t.p1-t.p0
        v3 = p-t.p0

        # Compute dot products
        dot00 = dot(v0, v0)
        dot01 = dot(v0, v1)
        dot02 = dot(v0, v2)
        dot11 = dot(v1, v1)
        dot12 = dot(v1, v2)

        # Compute barycentric coordinates
        inv_denom = 1 / (dot00 * dot11 - dot01 * dot01)
        u = (dot11 * dot02 - dot01 * dot12) * inv_denom
        v = (dot00 * dot12 - dot01 * dot02) * inv_denom

        # Check if point is in triangle
        return (u >= 0) && (v >= 0) && (u + v < 1)
    end
    """
    intersect with the triangle, return the nearest positive point or +infinity
    """
    function intersect(triangle::Triangle, ray::Ray)
        n = cross(triangle.p0-triangle.p1, triangle.p0-triangle.p2)
        d = norm(triangle.p0)/dot(n, triangle.p0)
        dist = (d-dot(ray.origin, n))/dot(ray.origin, n)
        if dist<0 || !inside_triangle(triangle, ray.origin+dist*ray.dir)
            return Inf
        else
            return dist
        end
    end

end

File: tracer/test/geometry.cc
#include "catch.hpp"
#include "../src/geometry.h"
#include "../src/triangle.h"
#include "../src/plane.h"
#include "../src/sphere.h"
#include "../src/ray.h"
#include "../src/intersect.h"
#include <opencv2/opencv.hpp>
#include <utility>

TEST_CASE("test virtual", "[geometry]"){
    using namespace std;
    using namespace cv;

    Geometry* triangle = new Triangle(Vec3f(1.0, 0.0, 6.0), Vec3f(0.0, 1.0, 6.0), Vec3f(1.0, 1.0, 6.0));
    auto inter = Intersect();

    auto ray = Ray();
    Vec3f ux = Vec3f(1.0, 0.0, 0.0);
    Vec3f uy = Vec3f(0.0, 1.0, 0.0);
    Vec3f uz = Vec3f(0.0, 0.0, 1.0);

    ray.d = uz;
    ray.o = 0.5*ux+0.5*uy;
    triangle->intersect(ray, inter);
    REQUIRE(inter.dist == Approx(6.0));
    REQUIRE(inter.norm == -uz);
    REQUIRE(inter.pos == 0.5*(ux+uy)+6*uz);
}

TEST_CASE("test triangle", "[geometry]"){
    using namespace std;
    using namespace cv;

    float inf = std::numeric_limits<float>::infinity();
    auto inter = Intersect();
    Vec3f ux = Vec3f(1.0, 0.0, 0.0);
    Vec3f uy = Vec3f(0.0, 1.0, 0.0);
    Vec3f uz = Vec3f(0.0, 0.0, 1.0);

    Triangle triangle(Vec3f(1.0, 0.0, 6.0), Vec3f(0.0, 1.0, 6.0), Vec3f(1.0, 1.0, 6.0));
    SECTION("hit triangle"){
        auto ray = Ray();
        ray.d = uz;
        ray.o = 0.5*ux+0.5*uy;
        triangle.intersect(ray, inter);
        REQUIRE(inter.dist == Approx(6.0));
        REQUIRE(inter.norm == -uz);
        REQUIRE(inter.pos == 0.5*(ux+uy)+6*uz);
    }

    SECTION("miss triange"){
        auto ray = Ray();
        ray.d = uz;
        ray.o = 2*ux+2*uy;
        triangle.intersect(ray, inter);
        REQUIRE(inter.dist == inf);
    }
}

TEST_CASE("test sphere", "[geometry]"){

    using namespace std;
    using namespace cv;

    float inf = std::numeric_limits<float>::infinity();
    auto inter = Intersect();
    Vec3f ux = Vec3f(1.0, 0.0, 0.0);
    Vec3f uy = Vec3f(0.0, 1.0, 0.0);
    Vec3f uz = Vec3f(0.0, 0.0, 1.0);

    Sphere sphere(6*uz, 2);
    SECTION("hit the sphere"){
        auto ray = Ray();
        ray.d = uz;
        ray.o = 0*uz;
        sphere.intersect(ray, inter);
        REQUIRE(inter.dist == Approx(6-2));
        REQUIRE(inter.norm == -uz);
        REQUIRE(inter.pos == 4*uz);
    }
    SECTION("miss the sphere"){
        auto ray = Ray();
        ray.d = uz;
        ray.o = 1.5*ux+1.5*uy;
        sphere.intersect(ray, inter);
        REQUIRE(inter.dist == inf);
    }

}

